api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java

```diff
@@ -495,17 +495,24 @@ private FirecloudWorkspaceId generateFirecloudWorkspaceId(String namespace, Stri
 
   @Override
   public ResponseEntity<WorkspaceResponseListResponse> getWorkspaces() {
-    // TODO: use FireCloud to determine what workspaces to return, instead of just returning
-    // workspaces from our database.
+    org.pmiops.workbench.db.model.Workspace dbWorkspace;
+    List<org.pmiops.workbench.firecloud.model.WorkspaceResponse> fcWorkspaces;
     User user = userProvider.get();
     List<WorkspaceResponse> responseList = new ArrayList<WorkspaceResponse>();
     if (user != null) {
-      for (WorkspaceUserRole userRole : user.getWorkspaceUserRoles()) {
-        // TODO: Use FireCloud to determine access roles, not our DB
+      try {
+        fcWorkspaces = fireCloudService.getWorkspaces();
+      } catch (org.pmiops.workbench.firecloud.ApiException e) {
+        throw ExceptionUtils.convertFirecloudException(e);
+      }
+      for ( org.pmiops.workbench.firecloud.model.WorkspaceResponse fcWorkspace : fcWorkspaces) {
         WorkspaceResponse currentWorkspace = new WorkspaceResponse();
-        currentWorkspace.setWorkspace(TO_CLIENT_WORKSPACE.apply(userRole.getWorkspace()));
-        currentWorkspace.setAccessLevel(userRole.getRole());
-        responseList.add(currentWorkspace);
+        dbWorkspace = workspaceService.get(fcWorkspace.getWorkspace().getNamespace(), fcWorkspace.getWorkspace().getName());
```

Comment:
Is this going to make a DB query for each workspace in the list? If so, can we potentially do one query?

Reply:
+1. We don't want to do O(n) DB queries.